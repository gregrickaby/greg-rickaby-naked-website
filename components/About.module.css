.about {
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  max-width: 740px;
  padding: 0 1.2rem;
}

@media (min-width: 768px) {
  .about {
    padding: 0;
  }
}
