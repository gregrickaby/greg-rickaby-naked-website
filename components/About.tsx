import styles from "./About.module.css";

export default function About() {
  return (
    <div id="about" className={styles.about}>
      <h2>About</h2>
      <p>
        Way back in 1997, I built my first website on Geocities and have been
        passionate about the web since! In my spare time I enjoy{" "}
        <a href="https://gregrickaby.smugmug.com" rel="author">
          taking photos
        </a>
        , cheering for the{" "}
        <a href="https://www.packers.com/">Green Bay Packers</a>, and making
        pizzas.
      </p>

      <h3>Web Developer</h3>
      <p>
        I&apos;m a Tech Lead at{" "}
        <a href="https://americaneagle.com/">AmericanEagle.com</a>, an industry
        leader in web design, development, hosting, and digital marketing. I
        partner with clients and team members to build websites with Next.js and
        WordPress.
      </p>
      <p>
        View my r√©sum√© on{" "}
        <a href="https://www.linkedin.com/in/gregrickaby/" rel="author">
          LinkedIn
        </a>
        .
      </p>

      <h3>Published Author</h3>
      <p>
        In 2017, I wrote a children&apos;s book titled{" "}
        <a href="https://amzn.to/34QAxAQ">
          <em>
            <strong>Creating a Website for Dummies Jr</strong>
          </em>
        </a>{" "}
        which was published by{" "}
        <a href="https://www.wiley.com/en-us"> Wiley Global</a> under the{" "}
        <a href="https://www.dummies.com/">dummies‚Ñ¢Ô∏è</a> brand.
      </p>

      <p>
        The book guides young readers through the planning and development
        processes of building a website. It has been translated into other 2
        other languages, and currently has a 4.5-star rating on Amazon.
      </p>

      <p>
        Since being published, I&apos;ve also served as the Technical Editor on
        three other books:
      </p>
      <ul>
        <li>
          <a href="https://amzn.to/37XCflU" rel="nofollow">
            <em>WordPress for Dummies</em>
          </a>{" "}
          (Sabin-Wilson, 2021)
        </li>
        <li>
          <a href="https://amzn.to/2BgCg7i" rel="nofollow">
            <em>Professional WordPress Plugin Development</em>
          </a>{" "}
          (Williams et al., 2020)
        </li>
        <li>
          <a href="https://amzn.to/37MMDLp" rel="nofollow">
            <em>WordPress All-In-One For Dummies</em>
          </a>{" "}
          (Sabin-Wilson, 2019)
        </li>
      </ul>

      <h3>Photographer</h3>
      <p>
        In late 2019, I bought a mirrorless camera and spent months taking
        courses, reading, practicing, and having my work critiqued by peers.
        What has started as an enjoyable hobby, has turned into a passionate way
        to express my creativity.
      </p>
      <p>
        View my{" "}
        <a href="https://gregrickaby.smugmug.com" rel="author">
          portfolio
        </a>{" "}
        on SmugMug.
      </p>

      <h3>Contributions</h3>
      <p>
        I've been part of the WordPress community since 2008. I've made
        contributions to core, documentation, plugins, and themes. I've also
        spoken at WordCamps and meet-ups. I'm also a contributor to other
        open-source projects like Next.js, Storybook, and Gatsby.
      </p>
      <p>
        Outside of tech, I volunteer my time and efforts at local community
        organizations. The Boy Scouts, Wiregrass Church, and the Enterprise High
        School Band Boosters to name a few.
      </p>

      <p>
        You can find me around the web with the links below. Stay safe out
        there, and thanks for reading!
        <span role="img" aria-label="cheers">
          üçª
        </span>
      </p>
    </div>
  );
}
