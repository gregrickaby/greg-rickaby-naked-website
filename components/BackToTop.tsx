'use client'

import {IconArrowUp} from '@tabler/icons-react'
import {useEffect, useState} from 'react'

/**
 * The back to top component.
 */
export default function BackToTop() {
  const [showButton, setShowButton] = useState(false)
  const buttonText = 'Go back to the top of the page'

  /**
   * Scroll handler.
   */
  const handleScroll = () => {
    window.scrollTo({
      top: 0,
      behavior: 'smooth'
    })
  }

  /**
   * Effect for showing the back to top button.
   */
  useEffect(() => {
    // Scroll event handler.
    const handleScroll = () => {
      setShowButton(window.scrollY > 200)
    }

    // Add event listener.
    window.addEventListener('scroll', handleScroll)

    // Cleanup event listener.
    return () => window.removeEventListener('scroll', handleScroll)
  }, [])

  return showButton ? (
    <button
      aria-label={buttonText}
      className="button fixed bottom-8 right-3 rounded-full bg-gray-400 p-2 text-white shadow-lg hover:bg-gray-500 lg:right-12 dark:bg-gray-800 hover:dark:bg-gray-700"
      data-testid="back-to-top"
      onClick={handleScroll}
      title={buttonText}
    >
      <IconArrowUp height="32" width="32" />
      <span className="sr-only">{buttonText}</span>
    </button>
  ) : null
}
