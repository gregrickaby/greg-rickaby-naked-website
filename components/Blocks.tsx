'use client'

import '@fancyapps/ui/dist/fancybox/fancybox.css'
import {useEffect, useRef} from 'react'

/**
 * The blocks component.
 */
export default function Blocks({content}: Readonly<{content: string}>) {
  // Create a ref for the post content.
  const contentRef = useRef<HTMLDivElement>(null)

  /**
   * Initialize photo gallery.
   */
  async function galleryInit() {
    // On the server? Bail.
    if (typeof window === 'undefined') {
      return
    }

    // No ref? Bail.
    if (!contentRef.current) {
      return
    }

    // Attempt to find the gallery.
    const gallery = contentRef.current.querySelector('.grd-photo-gallery-grid')

    // No gallery? Bail.
    if (!gallery) {
      return
    }

    // Import the libraries on demand.
    const Masonry = (await import('masonry-layout')).default
    const Fancybox = (await import('@fancyapps/ui')).Fancybox
    const imagesLoaded = (await import('imagesloaded')).default

    // Initialize ImagesLoaded.
    const imagesLoadedInstance = imagesLoaded(gallery)

    // Initialize Masonry.
    const masonryGrid = new Masonry(gallery, {
      gutter: 24,
      itemSelector: '.grd-photo-gallery-image',
      columnWidth: '.grd-photo-gallery-grid-sizer',
      percentPosition: true
    })

    // When ImagesLoaded is done, initialize Masonry.
    imagesLoadedInstance.on('done', () => {
      masonryGrid?.layout?.()
    })

    // Initialize Fancybox.
    Fancybox.bind('[data-fancybox]', {
      animated: false,
      groupAll: true,
      defaultDisplay: 'flex',
      Toolbar: {
        display: {
          left: ['infobar'],
          middle: [
            'zoomIn',
            'zoomOut',
            'toggle1to1',
            'rotateCCW',
            'rotateCW',
            'flipX',
            'flipY'
          ],
          right: ['slideshow', 'thumbs', 'download', 'close']
        }
      }
    })

    // Destroy Masonry and Fancybox when the component unmounts.
    return () => {
      masonryGrid?.destroy?.()
      Fancybox.destroy?.()
    }
  }

  /**
   * Initialize the gallery when the component mounts.
   */
  useEffect(() => {
    galleryInit()
  }, [content])

  return (
    <div
      ref={contentRef}
      className="text-lg"
      dangerouslySetInnerHTML={{__html: content}}
    />
  )
}
