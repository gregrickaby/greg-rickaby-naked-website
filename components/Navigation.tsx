import config from '@/lib/config'
import Link from 'next/link'

/**
 * Navigation component.
 */
export default function Navigation() {
  return (
    <nav
      className="flex flex-wrap items-center justify-center gap-4 md:justify-start"
      data-testid="nav"
    >
      {config.navigation.map((item) => (
        <Link className="underline" key={item.name} href={item.url}>
          {item.name}
        </Link>
      ))}
    </nav>
  )
}
