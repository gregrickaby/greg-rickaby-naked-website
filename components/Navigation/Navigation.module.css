.nav {
  @apply relative;
}

.navContainer {
  @apply flex items-center justify-between;
}

.navButton {
  @apply flex lg:hidden;
}

.linkContainer {
  @apply absolute right-0 top-12 z-10 text-right;
  @apply flex-col gap-4 rounded-lg pb-8 pl-8 pr-0 pt-8;
  @apply lg:static lg:flex lg:flex-row lg:rounded-none lg:bg-transparent lg:shadow-none;
  @apply bg-white dark:bg-gray-950;
}

.link {
  @apply text-xl hover:underline;
}
