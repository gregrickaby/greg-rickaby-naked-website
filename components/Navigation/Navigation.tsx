'use client'

import config from '@/lib/config'
import {IconMenu, IconX} from '@tabler/icons-react'
import clsx from 'clsx'
import Link from 'next/link'
import {MouseEvent, useEffect, useRef, useState} from 'react'
import styles from './Navigation.module.css'

/**
 * Navigation component.
 */
export function Navigation() {
  const [isOpen, setIsOpen] = useState(false)
  const navRef = useRef<HTMLDivElement | null>(null)

  /**
   * Handle link click.
   */
  const handleLinkClick = () => {
    setIsOpen(false)
  }

  /**
   * Handle click outside.
   *
   * @param event - Mouse event.
   * @returns void
   */
  const handleClickOutside = (event: MouseEvent | Event) => {
    if (navRef.current && !navRef.current.contains(event.target as Node)) {
      setIsOpen(false)
    }
  }

  /**
   * Add event listener.
   */
  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [])

  return (
    <nav ref={navRef} className={styles.nav} data-testid="nav">
      <div className={styles.navContainer}>
        <button
          aria-label="Toggle navigation"
          className={clsx('button', styles.navButton)}
          onClick={() => setIsOpen(!isOpen)}
        >
          {isOpen ? <IconX size={24} /> : <IconMenu size={24} />}
        </button>
      </div>
      <div
        className={clsx(styles.linkContainer, {flex: isOpen, hidden: !isOpen})}
        data-testid="nav-links"
      >
        {config.navigation.map((item) => (
          <Link
            className={styles.link}
            href={item.url}
            key={item.name}
            onClick={handleLinkClick}
          >
            {item.name}
          </Link>
        ))}
      </div>
    </nav>
  )
}
