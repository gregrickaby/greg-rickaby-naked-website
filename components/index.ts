export * from './ArticleCard'
export * from './BackToTop'
export * from './Blocks'
export * from './BlogArchive'
export * from './Byline'
export * from './CommentForm'
export * from './Comments'
export * from './Footer'
export * from './Header'
export * from './JsonLd'
export * from './Navigation'
export * from './SocialLinks'
