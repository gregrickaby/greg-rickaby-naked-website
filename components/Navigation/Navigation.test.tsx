import {Navigation} from '@/components/Navigation'
import config from '@/lib/config'
import '@testing-library/jest-dom'
import {act, fireEvent, render, screen} from '@testing-library/react'
import {axe} from 'jest-axe'
import {describe, expect, it} from 'vitest'

describe('Navigation', () => {
  it('should render', () => {
    const {container} = render(<Navigation />)
    expect(container).toBeInTheDocument()
  })

  it('should render navigation links', () => {
    render(<Navigation />)
    const button = screen.getByRole('button')

    // Open the menu.
    act(() => {
      fireEvent.click(button)
    })

    // Get all links.
    const links = screen.getAllByRole('link')

    // Verify the number of links.
    expect(links.length).toBe(config.navigation.length)

    // Verify each link is in the DOM.
    links.forEach((link) => {
      expect(link).toBeInTheDocument()
    })
  })

  it('should toggle the hamburger menu', () => {
    render(<Navigation />)
    const button = screen.getByRole('button')

    // Open the menu.
    act(() => {
      fireEvent.click(button)
    })

    // Get the navigation links container.
    const navLinks = screen.getByTestId('nav-links')

    // Verify the navigation links are visible.
    expect(navLinks).toHaveClass(/flex/i)

    // Close the menu.
    act(() => {
      fireEvent.click(button)
    })

    // Verify the navigation links are hidden.
    expect(navLinks).toHaveClass(/hidden/i)
  })

  it('should close the menu when clicking on a link', () => {
    render(<Navigation />)
    const button = screen.getByRole('button')

    // Open the menu.
    act(() => {
      fireEvent.click(button)
    })

    // Get all links.
    const links = screen.getAllByRole('link')

    // Click the first link.
    act(() => {
      fireEvent.click(links[0])
    })

    // Get the navigation links container.
    const navLinks = screen.getByTestId('nav-links')

    // Verify the navigation links are hidden.
    expect(navLinks).toHaveClass(/hidden/i)
  })

  it('should close the menu when clicking outside', () => {
    render(<Navigation />)
    const button = screen.getByRole('button')

    // Open the menu.
    act(() => {
      fireEvent.click(button)
    })

    // Click outside the menu.
    act(() => {
      fireEvent.mouseDown(document)
    })

    // Get the navigation links container.
    const navLinks = screen.getByTestId('nav-links')

    // Verify the navigation links are hidden.
    expect(navLinks).toHaveClass(/hidden/i)
  })

  it('should not have any accessibility issues', async () => {
    const {container, getByRole} = render(<Navigation />)
    const button = getByRole('button')

    // Open the menu.
    await act(async () => {
      fireEvent.click(button)
    })

    // Get the container.
    const results = await act(async () => {
      return await axe(container)
    })

    // Verify there are no violations.
    expect(results).toHaveNoViolations()
  })

  it('should match snapshot', () => {
    const {getByTestId} = render(<Navigation />)
    const nav = getByTestId('nav')
    expect(nav).toMatchSnapshot()
  })
})
