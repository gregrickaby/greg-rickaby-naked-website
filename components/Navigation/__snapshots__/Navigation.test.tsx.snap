// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation > should match snapshot 1`] = `
<nav
  class="_nav_2c0fa0"
  data-testid="nav"
>
  <div
    class="_navContainer_2c0fa0"
  >
    <button
      aria-label="Toggle navigation"
      class="button _navButton_2c0fa0"
    >
      <svg
        class="tabler-icon tabler-icon-menu "
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 8l16 0"
        />
        <path
          d="M4 16l16 0"
        />
      </svg>
    </button>
  </div>
  <div
    class="_linkContainer_2c0fa0 hidden"
    data-testid="nav-links"
  >
    <a
      class="_link_2c0fa0"
      href="/about"
    >
      About
    </a>
    <a
      class="_link_2c0fa0"
      href="/blog"
    >
      Blog
    </a>
    <a
      class="_link_2c0fa0"
      href="/contact"
    >
      Contact
    </a>
    <a
      class="_link_2c0fa0"
      href="/cv"
    >
      CV
    </a>
    <a
      class="_link_2c0fa0"
      href="/media"
    >
      Media
    </a>
    <a
      class="_link_2c0fa0"
      href="https://www.threads.net/@gregoryrickaby"
    >
      Photos
    </a>
    <a
      class="_link_2c0fa0"
      href="/resources"
    >
      Resources
    </a>
    <a
      class="_link_2c0fa0"
      href="/feed.xml"
    >
      RSS
    </a>
  </div>
</nav>
`;
